extension/src/lib.rs

```diff
@@ -0,0 +1,295 @@
+
+use std::{
+    collections::VecDeque,
+    mem::size_of,
+};
+
+use timescale_extension_utils::{
+    datum::ToDatum,
+    elog,
+    pg_agg,
+    pg_sys::{
+        ArrayType,
+        Datum,
+        FLOAT8OID,
+        FLOAT8PASSBYVAL,
+        TimestampTz,
+        construct_md_array,
+    },
+};
+
+type Milliseconds = i64;
+const USECS_PER_SEC: i64 = 1_000_000;
+const USECS_PER_MS: i64 = 1_000;
+
+pg_agg!{
+    // prom divides time into no-sliding windows of fixed size, e.g.
+    // |  5 seconds  |  5 seconds  |  5 seconds  |  5 seconds  |  5 seconds  |
+    // we take the first and last values in that bucket and uses `last-first` as the
+    // value for that bucket.
+    //  | a b c d e | f g h i | j   k |   m    |
+    //  |   e - a   |  i - f  | k - j | <null> |
+    pub fn gapfill_delta_transition(
+        state: Option<Pox<GapfillDeltaTransition>>,
+        lowest_time: TimestampTz,
+        greatest_time: TimestampTz,
+        step_size: Milliseconds, // `prev_now - step` is where the next window starts
+        window_size: Milliseconds, // the size of a window to delta over
+        time: TimestampTz,
+        val: f64,
+    ) -> Option<Pox<GapfillDeltaTransition>> {
+        if time < lowest_time || time > greatest_time {
+            elog!(Error, "input time less than lowest time")
+        }
+
+        let mut state = state.unwrap_or_else(|| {
+            let state: Pox<_> = GapfillDeltaTransition::new(
+                lowest_time,
+                greatest_time,
+                window_size,
+                step_size,
+                false,
+                false,
+            ).into();
```

Comment:
what's into again?

Reply:
arbitrary lossless type-conversion, in this case `Pox`ing the value